core/src/main/java/org/infinispan/interceptors/impl/PassivationCacheLoaderInterceptor.java

```diff
@@ -18,47 +18,52 @@
    @Inject ActivationManager activationManager;
 
    @Override
    public CompletionStage<InternalCacheEntry<K, V>> loadAndStoreInDataContainer(InvocationContext ctx, Object key,
                                                                                 int segment, FlagAffectedCommand cmd) {
-      Supplier<CompletionStage<InternalCacheEntry<K, V>>> supplier = () -> super.loadAndStoreInDataContainer(ctx, key, segment, cmd);
-      return handlePassivationLoad(key, segment, orderer, activationManager, supplier);
-   }
-
-   static <K, V> CompletionStage<InternalCacheEntry<K, V>> handlePassivationLoad(Object key, int segment,
-                                                                                 DataOperationOrderer orderer,
-                                                                                 ActivationManager activationManager,
-                                                                                 Supplier<CompletionStage<InternalCacheEntry<K, V>>> supplier) {
       CompletableFuture<Operation> future = new CompletableFuture<>();
       CompletionStage<Operation> delayStage = orderer.orderOn(key, future);
 
       CompletionStage<InternalCacheEntry<K, V>> retrievalStage;
-      if (delayStage != null) {
-         retrievalStage = delayStage.thenCompose(ignore -> supplier.get());
+      if (delayStage != null && !CompletionStages.isCompletedSuccessfully(delayStage)) {
+         retrievalStage = delayStage.thenCompose(ignore -> super.loadAndStoreInDataContainer(ctx, key, segment, cmd));
       } else {
-         retrievalStage = supplier.get();
+         retrievalStage = super.loadAndStoreInDataContainer(ctx, key, segment, cmd);
       }
-      return retrievalStage.whenComplete((value, t) -> {
-         if (value != null) {
-            if (log.isTraceEnabled()) {
-               log.tracef("Activating key: %s - not waiting for response", value.getKey());
-            }
-            // Note we don't wait on this to be removed, which allows the load to continue ahead. However
-            // we can't release the orderer acquisition until the remove is complete
-            activationManager.activateAsync(value.getKey(), segment)
-                  .whenComplete((ignore, throwable) -> {
-                     if (throwable != null) {
-                        log.warnf("Activation of key %s failed for some reason", t);
-                     }
-                     orderer.completeOperation(key, future, Operation.READ);
-                  });
+      if (CompletionStages.isCompletedSuccessfully(retrievalStage)) {
+         InternalCacheEntry<K, V> ice = CompletionStages.join(retrievalStage);
+         activateAfterLoad(key, segment, orderer, activationManager, future, ice, null);
+         return retrievalStage;
+      } else {
+         return retrievalStage.whenComplete((value, t) -> {
+            activateAfterLoad(key, segment, orderer, activationManager, future, value, t);
+         });
```

Comment:
nit:
```suggestion
         return retrievalStage.whenComplete((value, t) -> activateAfterLoad(key, segment, orderer, activationManager, future, value, t));
```

Reply:
I don't like that formatting, the line becomes way too long.